go/vt/vttablet/tabletmanager/shard_sync.go

```diff
@@ -236,23 +240,30 @@ func (agent *ActionAgent) startShardSync() {
 }
 
 func (agent *ActionAgent) stopShardSync() {
-	if agent.shardSyncCancel != nil {
-		agent.shardSyncCancel()
-		agent.shardSyncCancel = nil
-		agent.shardSyncChan = nil
+	agent.mutex.Lock()
+	if agent._shardSyncCancel != nil {
+		agent.mutex.Unlock()
+		agent._shardSyncCancel()
+		agent.mutex.Lock()
+		agent._shardSyncCancel = nil
+		agent._shardSyncChan = nil
 	}
+	agent.mutex.Unlock()
 }
 
 func (agent *ActionAgent) notifyShardSync() {
 	// If this is called before the shard sync is started, do nothing.
-	if agent.shardSyncChan == nil {
+	agent.mutex.Lock()
+	if agent._shardSyncChan == nil {
+		agent.mutex.Unlock()
 		return
 	}
+	agent.mutex.Unlock()
 
 	// Try to send. If the channel buffer is full, it means a notification is
 	// already pending, so we don't need to do anything.
 	select {
-	case agent.shardSyncChan <- struct{}{}:
+	case agent._shardSyncChan <- struct{}{}:
```

Comment:
This is reading `_shardSyncChan` again outside the lock. I think we should just `defer Unlock()` at the top and hold it until return.